### DISCUSSION TITLE
Add more (default) methods to problem class interface

### DISCUSSION THREAD

**Discussion by tlunet** [Parallel-in-Time/pySDC#253] [Ideas]
Following a _post by @brownbaerchen in https://github.com/Parallel-in-Time/pySDC/issues/244#issuecomment-1384938989_ , the current interface for `Problem` class in pySDC is as follow :

```python
class ptype(object):
    def __init__(self, init, dtype_u, dtype_f):
        # ...
     def eval_f(self, u, t):
        # ABSTRACT
    def apply_mass_matrix(self, u):
        # ABSTRACT
```

Looking at different current problem implementations, here are some ideas 

- define by default `apply_mass_matrix(self, u)` as an identity operator on `u` (since most problem don't have a mass matrix) and suggest to redefine it in a children class using, for instance, FEM.
- add `u_exact(self, t, u_init=None, t_init=None)` as a standard method using the [scipy reference solution for IVP](https://github.com/brownbaerchen/pySDC/blob/070d16e2a6f84ca21239b7b99b911c1c4c70bc2d/pySDC/implementations/problem_classes/Van_der_Pol_implicit.py#L75), and allow it to be redefined for problem that have a simpler exact solution
- add `solve_system(self, rhs, factor, u0, t)` as a base method to the problem class, with a generic implementation of a solution of the (non)-linear system `u-factor*f(u,t) = rhs` using the (very generic) [Newton-Krylov solver of scipy](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.newton_krylov.html), allow it to be redefined for problems that have a simpler way to solve this (_e.g_ for linear problems) or even for problem that don't allow any implicit solve (and use a custom ImplicitNotAllowedError).

:scroll: The change on `apply_mass_matrix` and `solve_system` would facilitate the binding with dedalus V3, as those are two main components required by dedalus time-integrators ...

Any comment/suggestion on that @pancetta @brownbaerchen ?

**Comment by pancetta** [comment]
When we first added FEniCS we needed to rewrite the sweeper and transfer classes, because now we had a mass matrix. If you don't have a mass matrix, those classes are easier to read and digest (and much closer to the literature). This is why we decided against a default `apply_mass_matrix` function. So far, only FEM needs that and this is what we're doing only 1% of the time. But maybe we should actually do that more often anyway?

**Reply by tlunet** [reply]
So far from what I saw in dedalus, they also have many situations where the mass matrix is actually identity. But their implementation solves generic PDE problem of the form (or something like)

```math
M \frac{dU}{dt} = L U + f(U,t)
```

with $L$ for the linear term and `f(U,t)` for the non linear term (solved with IMEX). Considering $M=I$ and $L=0$ allows to retrieve the classical form of ODE systems. Maybe a similar approach could be considered in pySDC, such that the problem interface itself allows complex ODE system formulations, but the default implementation sticks to the classical ODE representation ? That is :

```math
\frac{dU}{dt} = f(U,t)
```

**Reply by pancetta** [reply]
Sure, makes sense. One additional case where a mass matrix is needed is for compact FD stencils.. so, yes, maybe it's about time we don't treat this as an edge case but rather as a more generic way to work with PDE discretizations.

**Comment by tlunet** [comment]
After thinking about it on a (long ...) train travel, here are some propositions.

### Mass-matrix

Add it by default to define the problem
```math
M\frac{dU}{dt} = f(U,t)
```
and define a base method `applyMassMatrix(u)` that would simply return `u` in the base class. If someone wants a non-identity mass-matrix (_e.g_ for FEM or compact FD), its should simply be redefined in the specialized problem class. It also means that the generic sweeper should later use this `applyMassMatrix` method by default.

### IMEX 

Don't split $f(U,t)$ into non-linear and linear terms, for instance 
```math
M \frac{dU}{dt} = L U + F(U,t)
```
since some application may consider applying implicit sweep to a non-linear term, or a time-dependent $L(t)$ matrix. 
But keep the current generic ODE form with mass matrix (_i.e_ generic DAE form) and use the current idea of the `data_f` type, that can contains an `impl` and `expl` attribute to separate both IMEX part. Maybe a base class should be defined for pySDC datatypes in core module, with the IMEX datatype class (inheriting from base datatype) also defined in this core module. 

Additionally, the IMEX sweep could be added to the core modules, close to the base sweeper class that would contains both explicit and implicit implementations (and just testing the QDelta matrix to choose which one to use). That way, for one given application, only datatype and problem classes implementation is required, and the rest is only the correct choice of core classes and parameters already provided by pySDC core.

### Generic solver for implicit sweep 

Add a generic `solveSystem(rhs, factor, u0, t)` that use the Krylov-Newton solver from Scipy to solve
```math
U - factor*f(U,t) = rhs
```
with `u0` as starting solution for the iterative solver. This method could be redefined for linear problems or when the Jacobian matrix of `f(U,t)` is available and allow a more efficient solve (_e.g_ for the Lorenz system).

### Generic ODE solver for `u_exact`

Define a base generic method `uExact(t, u0=None, t0=0)` which would use the Scipy approach from @brownbaerchen to solve the ODE system when the `applyMassMatrix()` method is not redefined (quick easy test to do in Python). If `u0 is None`, uses the initial solution associated to the problem, given by a `u0` property set as abstract in the base class, that must be given with each problem implementation.

For cases when the applyMassMatrix function is redefined, use a more complex approach for generic DAE (see for instance this [stackoverflow post](https://stackoverflow.com/questions/23578596/solve-an-implicit-ode-differential-algebraic-equation-dae)).

Would you validate/correct that @pancetta ?

**Comment by pancetta** [comment]
Thanks a lot @tlunet! Let me try to comment one by one:
* Mass-matrix
Yes! We have to remember that this changes the base transfer, too.

* IMEX
pySDC does not enforce linear/nonlinear splitting. It's up to the user to define what `f.expl`, `f.impl` or `solve_system` means. I'd rather not move the sweeper implementation to the core classes, this is a clear implementation thing. We can think about setting defaults, yes, but from the implementation classes.

*  Generic solver for implicit sweep
Interesting thought, I like it!

* Generic ODE solver for `u_exact`
First part: yes. Second part: We're currently adding DAE support to pySDC (through a Master's project) and I don't want to interfere here. Also, I think only the most simplest things should be set by default and if users diverge, they need to take care of the implications themselves. If it turns out to be necessary (depending how people make use of the mass matrix feature), we can still reconsider.

**Comment by brownbaerchen** [comment]
I have no idea about mass matrices, so I can't give any comments on that.

I do however like the idea of implementing IMEX on a more basic level. In particular, I would like to have a function like
```
eval_f_fully_implicit
```
which would in the case of IMEX return the sum of the two rhs evaluations and in the implicit sweeper just the implicit part and so on.
Since I never used any other sweeper, I don't know if this can be generalised to those as well.

This would allow to have a bit more abstraction in the sweeper classes. For instance, you wouldn't need to overload the [`integrate` function](https://github.com/Parallel-in-Time/pySDC/blob/ecbccf931ce6a6c261969c08b259791a84c3a337/pySDC/implementations/sweeper_classes/imex_1st_order.py#L37) in the IMEX sweeper anymore.

The other place where this is useful is the scipy stuff. There you need to supply something that evaluates the full rhs. Ofc naming it "fully_implicit" may not make sense, since in this case "fully_explicit" is probably closer to the truth, but whatever.

We can also just implement a function like this for the scipy reference solution stuff, if @pancetta wants to keep the non-imex sweepers a bit more simple, though.

**Reply by tlunet** [reply]
I do like this idea of full rhs evaluation, but I wonder if this should be done rather in the `data_f` class than the problem class. 

I would imagine that the base `data_f` class must have a `vals` attribute that would expose data values in the generic case, _e.g_ the numpy array for `mesh` datatype, and for IMEX datatype it would simply sum the implicit and explicit part and return it.

Eventually, this could also be a property that would prevent writing on it, except if you are in the problem class, and then fully aware of what the `data_f` type is, and then using the specialized data setter (e.g `f.vals[:,:,:] = ...` for a 3D problem)

**Reply by pancetta** [reply]
I spend a lot of time getting rid of a dedicated attribute (called `values` in previous versions, pre-v4), because then you need to write functions to do stuff like `add`, `rmult` and so on (or override those methos). This usually involves instantiating new data types each time you do that and this is very costly. If you can avoid this, fine, if not, I won't go back to dedicated attributes.

**Reply by pancetta** [reply]
The way it's currently done/preferred is to subclass an existing datatype (e.g. `ndarray`) and modify the `__new__` function. This way all operations are preserved and no further overhead is added.

**Reply by tlunet** [reply]
Should I consider then that the `fenics_mesh` class is supposed to be deprecated ?

**Reply by pancetta** [reply]
Not all data structures can be subclassed. The one from FEniCS are tricky. I'm trying to get FEniCSx to work, there I'd probably rely on `mesh` as a wrapper. So, no, not deprecated.

**Comment by tlunet** [comment]
So ... it appears that implementing those generic methods for problem classes need also a rework of the datatype classes, with the introduction of a base interface for `u` and a base generic class for `evalF` output, that would be directly linked to the datatype of `u`. 

Here is a first draft that tries to solve those issues : https://github.com/Parallel-in-Time/pySDC/blob/847e8ca1491df1076aff9c61768245ff940101a9/pySDC/playgrounds/datatypes/draft.py. I'm slowly checking how the current class could be translated and used into this new structure. Let me know if you see any issue and supplementary test I could try.

@pancetta : for some reason I don't understand yet, it seems that we don't need the complicated approach to subclass `ndarray`, as described [here ...](https://numpy.org/doc/stable/user/basics.subclassing.html)

**Reply by pancetta** [reply]
The `clone` function is precisely what I would like to avoid. You call this each time to do a `+` and so on and this is very costly!

**Reply by pancetta** [reply]
Rewriting the `__new__` function is not difficult and very short. Yes, it does not always work, but if it does, it'll reduce boilerplate code.

**Reply by pancetta** [reply]
Also, think about CUDA code.. here creating a data structure is something to avoid at all costs.

**Reply by tlunet** [reply]
> The `clone` function is precisely what I would like to avoid. You call this each time to do a `+` and so on and this is very costly!

Yes, but you do it implicitly too with `ndarray`, _e.g_ :

```python
L.dt * self.coll.Qmat[m, 1] * (L.f[1].impl + L.f[1].expl)
```

- create a new `ndarray` to store `L.f[1].impl + L.f[1].expl)`
- create a new `ndarray` to store `L.dt * self.coll.Qmat[m, 1] * (L.f[1].impl + L.f[1].expl)`

With the proposed structure, you do it one time for `DataTypeF` once you do one arithmetic operation with it, and then it's back to a subclass of `ndarray` that won't use clone by itself. So I would say that actually, `L.dt * self.coll.Qmat[m, 1] * (L.f[1])` would require only the creation of one `DataType` object, and then all modification would be done without copy (implicit and explicit part are added without additional copy using `toDataType()`, and multiplication is done without new copy in `__mul__`).

PS : the `copy=False` option for clone would also allow to avoid copy, as it is done yet in `solveSystem` and `uExact`

PPS : we don't want to do arithmetic operations between DataTypeF types, right ?

**Reply by tlunet** [reply]
> Rewriting the `__new__` function is not difficult and very short. Yes, it does not always work, but if it does, it'll reduce boilerplate code.

But this is required only to store the `_comm` attribute, that should be given to every datatype class when we create a new one (in particular, during arithmetic operations ...). Using it as a class attribute allows to avoid this, store the communicator in the same location for each datatype class, and allows to modify it during run-time (for instance, if we consider space-time parallelization that is initialized at the beginning).

And bonus : you don't need to rewrite every time the `isend`, `irecv`, ... methods for each specialized datatype, as they only need to expose an array-like interface through the `values` property to use the inherited methods.

**Reply by tlunet** [reply]
> Also, think about CUDA code.. here creating a data structure is something to avoid at all costs.

I don't understand this part ... is it linked to the `cupy_mesh.py` implementation ?

**Reply by pancetta** [reply]
> > The `clone` function is precisely what I would like to avoid. You call this each time to do a `+` and so on and this is very costly!
> 
> Yes, but you do it implicitly too with `ndarray`, _e.g_ :
> 
> ```python
> L.dt * self.coll.Qmat[m, 1] * (L.f[1].impl + L.f[1].expl)
> ```
> 
>     * create a new `ndarray` to store `L.f[1].impl + L.f[1].expl)`
> 
>     * create a new `ndarray` to store `L.dt * self.coll.Qmat[m, 1] * (L.f[1].impl + L.f[1].expl)`
> 
> 
> With the proposed structure, you do it one time for `DataTypeF` once you do one arithmetic operation with it, and then it's back to a subclass of `ndarray` that won't use clone by itself. So I would say that actually, `L.dt * self.coll.Qmat[m, 1] * (L.f[1])` would require only the creation of one `DataType` object, and then all modification would be done without copy (implicit and explicit part are added without additional copy using `toDataType()`, and multiplication is done without new copy in `__mul__`).
> 
> PS : the `copy=False` option for clone would also allow to avoid copy, as it is done yet in `solveSystem` and `uExact`
> 
> PPS : we don't want to do arithmetic operations between DataTypeF types, right ?

But don't you to this twice in your approach? Once explicitly, once implicitly? 

In any case, I think we need to benchmark this, right?

**Reply by tlunet** [reply]
I don't think so : that's the whole approach with this `DataTypeF` class and its `toDataType()` method. But you're right : I'll write down some benchmark tests to compare with the current `mesh` dataType in pySDC.

**Reply by pancetta** [reply]
> > Rewriting the `__new__` function is not difficult and very short. Yes, it does not always work, but if it does, it'll reduce boilerplate code.
> 
> But this is required only to store the `_comm` attribute, that should be given to every datatype class when we create a new one (in particular, during arithmetic operations ...). Using it as a class attribute allows to avoid this, store the communicator in the same location for each datatype class, and allows to modify it during run-time (for instance, if we consider space-time parallelization that is initialized at the beginning).
> 
> And bonus : you don't need to rewrite every time the `isend`, `irecv`, ... methods for each specialized datatype, as they only need to expose an array-like interface through the `values` property to use the inherited methods.

First of all, you don't write data types very often. You do this once and then it's fine. So, I don't care how complex it is, as long as it is fast. Also, depending on the data type, you better should use what the datatype provides (if anything).

The `__new__` function is also used to distinguish between instantiating data with a given value or by copying stuff during instantiation. Yes, there are other ways to do that, but doing it in `__new__` is rather elegant.

**Reply by pancetta** [reply]
> > Also, think about CUDA code.. here creating a data structure is something to avoid at all costs.
> 
> I don't understand this part ... is it linked to the `cupy_mesh.py` implementation ?

Look at the previous approach for the CuPy data type: https://github.com/Parallel-in-Time/pySDC/commit/55fb233765abd0ab9ab53fd77f9140d748eb3d87. This is much slower than using `__new__`.

**Reply by pancetta** [reply]
> PPS : we don't want to do arithmetic operations between DataTypeF types, right ?

What do you mean? Adding  two different data types (e.g. `fenics_mesh` and `cupy_mesh`)? No. We don't want that.

**Reply by tlunet** [reply]
> > PPS : we don't want to do arithmetic operations between DataTypeF types, right ?
> 
> What do you mean? Adding two different data types (e.g. `fenics_mesh` and `cupy_mesh`)? No. We don't want that.

No, I mean adding the result of two `eval_f` evaluation, for instance. In general, for one given application, adding two variables that have the same `dtype_f` type ...

**Reply by pancetta** [reply]
I'm confused and pretty sure I don't understand what you're saying: `f1 + f2` is a common operation, e.g. during integration, no? So, yes, of course we need to add `f`-evaluations.

**Reply by tlunet** [reply]
I mean : you never `dtype_f` + `dtype_f`, but rather `dtype_f` + `dtype_u`, that you store into `dtype_u`, right ?

**Reply by pancetta** [reply]
Sure you do `dtype_f + dtype_f`! Computing `Qf` needs exactly that.

**Reply by pancetta** [reply]
Oh, I see... hm. Very good question...

**Comment by pancetta** [comment]
Can you please test parallel code, too? E.g. distributed arrays with a communicator, computing a norm or so.

**Comment by pancetta** [comment]
So, we need different communicators for different instances of data. One `u` could belong to this space communicator, another `u` to another one. Can you do that with your approach?

**Reply by tlunet** [reply]
Yes, you can redefine the class attribute for one given instance. Example 

```python
class Data:
     comm = 1

u1 = Data()  # => inherit from comm = 1
u2 = Data()
u2.comm = 2 # => u2 has comm=2, but u1 still has comm=1
```

I believe it also work to define one specific communicator for different children class ..

**Reply by tlunet** [reply]
But indeed : if you copy u2, you don't get `comm=2` anymore ...

**Reply by pancetta** [reply]
I remember me thinking about this and failing, long time ago.

**Reply by tlunet** [reply]
However, this could be taken care of in the `clone` method

**Reply by pancetta** [reply]
So, to me, adding the communicator to the instance is the better solution. More explicit, safer.

**Reply by tlunet** [reply]
Can you just point me to an example where two datatypes would require different communicators ?

**Reply by pancetta** [reply]
Hm…  I though for space-parallelism, but on one process, they would all get the same. Similarly for time-parallelism, right? So… maybe I’m wrong and this scenario does not exist after all. Hm…..

**Reply by tlunet** [reply]
I guess. For space-time parallelism (doesn't seems to be implemented in pySDC yet), you would need two communicators, but each datatype should have access to both (and the specific methods to send/recv data in space or time). With the current approach, you would need to add a second `comm` attribute for each implemented datatypes, and the specific send/recv methods.

Using the proposed approach, that should be something done only once in the core `DataType` class, and all inherited datatypes would benefit from it.

**Reply by pancetta** [reply]
It surely is implemented, both with petsc4py and mpi4py-FFT. We tested this on 6k cores and I think this is even part of the CI.

**Reply by pancetta** [reply]
No second comm needed.

**Reply by tlunet** [reply]
Oh I see, because it's hidden in the `PETSc.Vec` object :

```python
# create DMDA object which will be used for all grid operations
da = PETSc.DMDA().create(
    [problem_params['cnvars'][0], problem_params['cnvars'][1]], stencil_width=1, comm=problem_params['comm']
) 
```

But then, I don't understand why the `mesh` datatype needs to store a time communicator for time parallelization, if the `petsc_vec` can actually use time parallelization without storing any ?

**Reply by pancetta** [reply]
Not time, but space communicators. mpi4py-fft needs this, see e.g. https://github.com/Parallel-in-Time/pySDC/blob/master/pySDC/implementations/problem_classes/AllenCahn_MPIFFT.py and https://github.com/Parallel-in-Time/pySDC/blob/master/pySDC/projects/Performance/run_simple_forcing_benchmark.py.

**Reply by pancetta** [reply]
The time communicator is done by the controller.

**Reply by tlunet** [reply]
Ah ok, thanks ! Got confused, since the communication methods for time are implemented within the datatypes, thought those were linked ...

**Reply by pancetta** [reply]
Ah, I see! So, the data types should know nothing about communication in time, but it has to tell the controller, how it can be communicated.

**Reply by tlunet** [reply]
That's what I realized after looking at the code in `controller_MPI`. For each datatypes, the communications method have to be redefined, but it's not explicit in some common interface.

With the proposed approach, this would be automatically set by the `DataType` base class. The developer (of a new inherited `DataType`) would simply have to specify with the `values` property (and setter) how to get array-like view on the data, and the base methods would do the rest. Eventually, for DataType that cannot be viewed into arrays, those methods could be redefined (but I don't see really a concrete example for this).

**Reply by pancetta** [reply]
Appealing II, I must admit. But again, we need to check for performance.

**Reply by tlunet** [reply]
Still ongoing ... but here are some preliminary results with one Forward Euler step. Dashed lines are the current implementation (reference) with `mesh` datatypes.
The `u + dt*f(u,t)` test consider both evaluation of `f` and multiplication, while the `u + dt*f` test only look at the FE update with the `f(u,t)` term already computed. See the current [test file](https://github.com/Parallel-in-Time/pySDC/blob/f6f698890fcc35d4451345a271096c80340f4390/pySDC/playgrounds/datatypes/check_performance.py) on `v6`.

![image](https://user-images.githubusercontent.com/12183839/214518105-1b6cbae6-119b-40a8-8549-4408d76e9bb8.png)

**Reply by pancetta** [reply]
So, you're telling me your approach may even be faster...?

**Reply by tlunet** [reply]
Not promising anything yet, but seems like it :smiley:

**Reply by pancetta** [reply]
We need to test this with PETSc and FEniCS, too. And I worry a bit about the GPU stuff...

**Comment by pancetta** [comment]
Bottom line: I don't see the benefit of rewriting the data types yet again. If the way you propose it faster, I'm on board. If not, I'd rather leave it like that, because we have so many open issues, even without touching the data types.

**Reply by tlunet** [reply]
This would allow :

- to write a generic `solveSystem` and `uExact` on the base level
- one unique implementation of `integrate` that would work for any datatypes
- one unique implementation of a generic sweeper (that would call `solveSystem` only for non-null diagonal `QDelta` coefficients),
- an easy definition of IMEX datatypes (or multicomponents for multi-implicit SDC) that could be combined with any main datatypes (mesh, petsc, fenics, particles, ...)
- ...

So mostly bonus things that may be not necessary yet, but could surely improve usage and development of pySDC. But I do agree : this must be adapted to the current datatype, and at least not slower than the current implementation (better if faster).

**Reply by pancetta** [reply]
Thanks! Don't me wrong, I really appreciate your effort and dedication!

> to write a generic solveSystem and uExact on the base level

This depends on the actual data type anyway, no? `solveSystem` is very different for FEniCS than for PETSc.

> one unique implementation of integrate that would work for any datatypes

Ok, well, but this is going to less explicit, right? In general the only dependency is on the number and type of `f` components. The rest (which specific type) is generic.

> one unique implementation of a generic sweeper (that would call solveSystem only for non-null diagonal QDelta coefficients)

Which sweepers would it replace? IMEX? implicit? explicit? Verlet? Multi-implicit?

> an easy definition of IMEX datatypes (or multicomponents for multi-implicit SDC) that could be combined with any main datatypes (mesh, petsc, fenics, particles, ...)

Appealing, indeed!
